pkg/controller/machine/machine.go

```diff
@@ -463,36 +459,58 @@ func (c *Controller) ensureMachineHasNodeReadyCondition(machine *clusterv1alpha1
 	})
 }
 
-func (c *Controller) cleanupMachineAfterDeletion(machine *clusterv1alpha1.Machine) error {
-	var err error
-	glog.V(4).Infof("Removing finalizers from machine machine %s", machine.Name)
+// deleteMachine makes sure that an instance has gone in a series of steps.
+func (c *Controller) deleteMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
+	if err := c.deleteCloudProviderInstance(prov, machine); err != nil {
+		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete instance at cloud provider: %v", err)
+		return err
+	}
 
-	if machine, err = c.updateMachine(machine, func(m *clusterv1alpha1.Machine) {
-		finalizers := sets.NewString(m.Finalizers...)
-		finalizers.Delete(finalizerDeleteInstance)
-		finalizers.Delete(finalizerDeleteNode)
-		m.Finalizers = finalizers.List()
-	}); err != nil {
-		return fmt.Errorf("failed to update machine after removing the delete instance finalizer: %v", err)
+	// Delete the node object after the instance is gone
+	if sets.NewString(machine.Finalizers...).Has(finalizerDeleteInstance) {
```

Comment:
IMHO we should first delete the node, before we delete the instance, especially since node deletion may imply draining

Reply:
The kubelet might restart & registers new - It's a safety mechanism to prevent empty node objects.

But I just looked it up in upstream, and the kubelet actually never fatals during runtime - it just loggs the error.
So unless any external entity restarts the kubelet on failed health-checks (And the gracetime period must be very, very short) we're safe.

I'll update the pull

Reply:
I changed the order in https://github.com/kubermatic/machine-controller/pull/348/commits/1e0d1d19356be90df3c0752b189fb404bd2a35bb

Reply:
Okay sorry for the back and forth but I just thought about this again. So if we delete the `node` beforehand and remove the finalizer there is a decent chance it will just re-register and stay, right?

Maybe the right order should be:

* Drain + mark node as unschedulable
* Delete the instance
* Delete the node

WDYT?

Reply:
The Kubelet only creates the Node object during startup - So the chance that the Node object appears again after it has been deleted is fairly small.

Deleting the instance before we remove the Node object will eliminate that chance though.
The above order sounds about right.